### PR TITLE
Improved and restructured gpu-draw API

### PR DESCRIPTION
**Pull request by emilwestergren**

### CODE REVIEW

File: source/draw/RasterImage.ooc [hunk 1]
```diff
@@ -78,4 +78,5 @@ RasterImage: abstract class extends Image {
 		}
 		result
 	}
+	save: virtual func (filename: String) -> Int { Debug raise("RasterImage save unimplemented for format!") }
```
Comment (davidhesselbom) [inline source/draw/RasterImage.ooc]: Nice!

Comment (davidhesselbom) [inline source/draw/RasterImage.ooc]: It should be possible to print the class name, too, but I think this will show in the backtrace anyway, so nevermind.

File: source/draw/gpu/GpuSurface.ooc [hunk 2]
```diff
@@ -1,29 +1,73 @@
...
-// (at your option) any later version.
-//
-// This program is distributed in the hope that it will be useful,
-// but WITHOUT ANY WARRANTY; without even the implied warranty of
-// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
-// GNU Lesser General Public License for more details.
-//
-// You should have received a copy of the GNU Lesser General Public License
-// along with this program. If not, see <http://www.gnu.org/licenses/>.
-
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuSurface.ooc]: Why was this removed?

File: source/draw/gpu/GpuSurface.ooc [hunk 3]
```diff
@@ -1,29 +1,73 @@
...
+		set(value) { this _view = this _toReference * value * this _toLocal }
+	}
+	_focalLength: Float
+	focalLength: Float {
+		get { this _focalLength }
+		set(value) {
+			this _focalLength = value
+			if (this _focalLength > 0.0f) {
+				this _projection = FloatTransform3D new(2.0f * this _focalLength / this size width, 0.0f, 0.0f, 0.0f, 0.0f, 2.0f * this _focalLength / this size height, 0.0f, 0.0f, 0.0f, 0.0f, -(this farPlane + this nearPlane) / (this farPlane - this nearPlane), -1.0f, 0.0f, 0.0f, -2.0f * this farPlane * this nearPlane / (this farPlane - this nearPlane), 0.0f)
+			}
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuSurface.ooc]: Unnecessary brackets

File: source/draw/gpu/GpuSurface.ooc [hunk 4]
```diff
@@ -1,29 +1,73 @@
...
+		get { this _toLocal * this _view * this _toReference }
+		set(value) { this _view = this _toReference * value * this _toLocal }
+	}
+	_focalLength: Float
+	focalLength: Float {
+		get { this _focalLength }
+		set(value) {
+			this _focalLength = value
+			if (this _focalLength > 0.0f) {
+				this _projection = FloatTransform3D new(2.0f * this _focalLength / this size width, 0.0f, 0.0f, 0.0f, 0.0f, 2.0f * this _focalLength / this size height, 0.0f, 0.0f, 0.0f, 0.0f, -(this farPlane + this nearPlane) / (this farPlane - this nearPlane), -1.0f, 0.0f, 0.0f, -2.0f * this farPlane * this nearPlane / (this farPlane - this nearPlane), 0.0f)
```
Comment (davidhesselbom) [inline source/draw/gpu/GpuSurface.ooc]: This is a REALLY long line.

File: source/draw/gpu/android/AndroidContext.ooc [hunk 5]
```diff
@@ -132,45 +138,27 @@ AndroidContext: class extends OpenGLES3Context {
 		target := this createYuv420Semiplanar(targetSize) as GpuYuv420Semiplanar
 		this _unpackRgbaToMonochrome targetSize = target y size
 		this _unpackRgbaToMonochrome sourceSize = source size
-		target y canvas draw(source, _unpackRgbaToMonochrome, IntBox2D new(target y size))
+		this _unpackRgbaToMonochrome transform = FloatTransform3D createScaling(source transform a, -source transform e, 1.0f)
+		//Must adapt to change in kean
+		//target y canvas draw(source, _unpackRgbaToMonochrome, IntBox2D new(target y size))
```
Comment (davidhesselbom) [inline source/draw/gpu/android/AndroidContext.ooc]: "Must adapt" - does that mean it's been adapted already, or is this a TODO for the future?

File: source/math/Quaternion.ooc [hunk 6]
```diff
@@ -279,9 +279,9 @@ Quaternion: cover {
 	sphericalLinearInterpolation: func(other: Quaternion, factor: Float) -> This {
 		cosAngle := this dotProduct(other)
 		longPath := cosAngle < 0.0f
-		angle := acos(Float absolute(cosAngle))
+		angle := acos(Float absolute(cosAngle) as Float clamp(-1.0f, 1.0f))
 		result: Quaternion
-		if (angle < 1e-10)
+		if (angle < 0.00000001)
```
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: Why not `1e-8`?

File: source/math/Quaternion.ooc [hunk 7]
```diff
@@ -313,6 +313,19 @@ Quaternion: cover {
 	relativeQuaternion: func(other: Quaternion) -> Quaternion {
 		other * this inverse
 	}
+	relativeFromVelocity: static func(angularVelocity: FloatPoint3D) -> This {
+		result := Quaternion identity
+		angle := sqrt(angularVelocity x * angularVelocity x + angularVelocity y * angularVelocity y + angularVelocity z * angularVelocity z)
+		if (angle > 0.00000001f) {
```
Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: Same here, `1e-8` should work and makes it a lot easier to count the zeroes.

File: source/draw/gpu/GpuSurface.ooc (hunk 2 above)
Comment (emilwestergren) [inline source/draw/gpu/GpuSurface.ooc]: Prob by mistake, readding.

File: source/draw/gpu/android/AndroidContext.ooc (hunk 5 above)
Comment (emilwestergren) [inline source/draw/gpu/android/AndroidContext.ooc]: It has been adapted but not well-tested. I'll add a TODO instead.

File: source/math/Quaternion.ooc (hunk 6 above)
Comment (emilwestergren) [inline source/math/Quaternion.ooc]: Didn't work (according to Niklas).

Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: But `1e-10` worked? That seems unlikely. What about `1e-08`, then?

File: source/math/Quaternion.ooc (hunk 7 above)
Comment (emilwestergren) [inline source/math/Quaternion.ooc]: Didn't work (according to Niklas).

File: source/math/Quaternion.ooc (hunk 6 above)
Comment (emilwestergren) [inline source/math/Quaternion.ooc]: None of them worked.

Comment (davidhesselbom) [inline source/math/Quaternion.ooc]: Oh, you're right, it has to be `1.0e-8` (or `1.0e-8f` for Floats).

Comment (emilwestergren) [inline source/math/Quaternion.ooc]: Works if you write 1.0e-10, so I will change it.

**Comment by emilwestergren** [comment]
Updated.

File: source/draw/RasterImage.ooc (hunk 1 above)
Comment (davidhesselbom) [inline source/draw/RasterImage.ooc]: Also, we need to return an `Int` here or we get compiler warnings.

File: source/draw/RasterYuv420Semiplanar.ooc [hunk 8]
```diff
@@ -166,7 +166,7 @@ RasterYuv420Semiplanar: class extends RasterYuvSemiplanar {
 		bgr free()
 		result
 	}
-	save: func (filename: String) {
+	save: override func (filename: String) -> Int {
```
Comment (davidhesselbom) [inline source/draw/RasterYuv420Semiplanar.ooc]: This also needs to return an `Int`, probably the result of the `bgr save(filename)` call.

File: source/draw/RasterUv.ooc [hunk 9]
```diff
@@ -125,7 +125,7 @@ RasterUv: class extends RasterPacked {
 		StbImage free(data)
 		This new(RasterBgr new(buffer, IntSize2D new (x, y)))
 	}
-	save: func (filename: String) -> Int {
+	save: override func (filename: String) -> Int {
```
Comment (davidhesselbom) [inline source/draw/RasterUv.ooc]: Unless `result` is freed, this will cause a memory leak.

**Comment by emilwestergren** [comment]
Updated.

**Comment by davidhesselbom** [comment]
Let's put this in the new upstream branch `gpu_cleanup` and fix what's broken before merging into `master`, shall we?
